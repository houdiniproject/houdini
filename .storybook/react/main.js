// License: LGPL-3.0-or-later
// Based on https://github.com/rails/webpacker/issues/1004#issuecomment-628377930
process.env.NODE_ENV = "development";
const railsWebpackEnv = require("../../config/webpack/environment");

module.exports = {
  stories: [
    "../../app/packs/**/*!(--html).stories.mdx",
    "../../app/packs/**/*!(--html).stories.[tj]s?(x)"
  ],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    'storybook-addon-intl'
  ],
  webpackFinal: (config) => {
    const result = {
    // do mutation to the config
    ...config,
    resolve: {
      ...config.resolve,
      ...railsWebpackEnv.config.resolve,
      modules: railsWebpackEnv.resolvedModules.map((i) => i.value),
    },
    module: {
      ...config.module,
      rules: railsWebpackEnv.loaders
        .filter((i) => !["nodeModules", //not sure why this is here
        "moduleCss" // this addresses issues with our webpack config for css not matching what storybook wants
      ].includes(i.key) )
        .map((i) => i.value),
    },
    plugins: [
      ...config.plugins,
      ...railsWebpackEnv.plugins.map((i) => i.value),
    ],
    
  };

  result.module.rules.filter((i) => i.test.test('.ttf')).forEach((i) => {
    i.use=['url-loader']
  })


  return result;
},
};
