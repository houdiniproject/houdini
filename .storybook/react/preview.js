// License: LGPL-3.0-or-later
import intlDecorate from '../../app/packs/components/tests/intl';
const jest = require('jest-mock');
window.jest = jest;

export const decorators = [intlDecorate()]