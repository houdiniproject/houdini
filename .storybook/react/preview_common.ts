// License: LGPL-3.0-or-or-later

import intlDecorate from '../../app/javascript/components/tests/intl';
import cssBaseline from '../../app/javascript/components/tests/decorators/baseline'
import clearSessionStorage from '../../app/javascript/components/tests/decorators/clearSessionStorage'
import { initialize, mswDecorator } from 'msw-storybook-addon';

initialize({
	onUnhandledRequest: 'bypass' // we need this because otherwise HMR and other parts of Storybook loading wouldn't work
});


export const decorators = [ clearSessionStorage(), mswDecorator, intlDecorate(), cssBaseline()]