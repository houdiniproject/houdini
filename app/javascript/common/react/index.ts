// License: LGPL-3.0-or-later

/**
 * Adds type safety to your default props.
 *
 * @example
 * // your component
 * function CustomComponent(props:{className:string, disabled?:boolean}) : JSX.Element {
 * 	// not show for
 * }
 *
 * CustomComponent.defaultProps = defaultProps(CustomComponent, {
 * 	className: 'another'
 * })
 *
 * @example <caption>Function properly typechecks</caption>
 * // use the same component as previous example
 * CustomComponent.defaultProps = defaultProps(CustomComponent, {
 * 	className: 'another',
 * 	unexpectedProp: 3 // <-- Typecheck will throw a compile error because you didn't expect this.
 * })
 *
 * @param {(props:TProperty) => JSX.Element} _component - the component you want to create default props for
 * @param {Partial<TProperty>} defaultProps - the defaultProps for your component
 * @returns defaultProps
 * */
export function defaultProps<TProperty>(_component:(props:TProperty) => JSX.Element, defaultProps:Partial<TProperty>):Partial<TProperty> {
	return defaultProps;
}