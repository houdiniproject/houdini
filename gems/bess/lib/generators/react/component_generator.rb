# frozen_string_literal: true

# License: AGPL-3.0-or-later WITH WTO-AP-3.0-or-later
# Full license explanation at https://github.com/houdiniproject/houdini/blob/master/LICENSE
# from: https://github.com/reactjs/react-rails/blob/master/lib/generators/react/component_generator.rb
module React
    module Generators
      class ComponentGenerator < ::Rails::Generators::NamedBase
        source_root File.expand_path '../../templates', __FILE__
        desc <<-DESC.strip_heredoc
        Description:
            Scaffold a React component into `components/` of your Webpacker source or asset pipeline.
            The generated component will include a basic render function and a PropTypes
            hash to help with development.
        Available field types:
            Basic prop types do not take any additional arguments. If you do not specify
            a prop type, the generic node will be used. The basic types available are:
            any
            array
            bool
            element
            func
            number
            object
            node
            shape
            string
            Special PropTypes take additional arguments in {}, and must be enclosed in
            single quotes to keep bash from expanding the arguments in {}.
            instanceOf
            takes an optional class name in the form of {className}
            oneOf
            behaves like an enum, and takes an optional list of strings that will
            be allowed in the form of 'name:oneOf{one,two,three}'.
            oneOfType.
            oneOfType takes an optional list of react and custom types in the form of
            'model:oneOfType{string,number,OtherType}'
        Examples:
            rails g react:component person name
            rails g react:component restaurant name:string rating:number owner:instanceOf{Person}
            rails g react:component food 'kind:oneOf{meat,cheese,vegetable}'
            rails g react:component events 'location:oneOfType{string,Restaurant}'
        DESC
  
        argument :attributes,
                 :type => :array,
                 :default => [],
                 :banner => 'field[:type] field[:type] ...'

        class_option :ts,
                     type: :boolean,
                     default: true,
                     desc: 'Output tsx class based component'

        REACT_PROP_TYPES = {
          'node' =>        'PropTypes.node',
          'bool' =>        'PropTypes.bool',
          'boolean' =>     'PropTypes.bool',
          'string' =>      'PropTypes.string',
          'number' =>      'PropTypes.number',
          'object' =>      'PropTypes.object',
          'array' =>       'PropTypes.array',
          'shape' =>       'PropTypes.shape({})',
          'element' =>     'PropTypes.element',
          'func' =>        'PropTypes.func',
          'function' =>    'PropTypes.func',
          'any' =>         'PropTypes.any',
  
          'instanceOf' => ->(type) {
            'PropTypes.instanceOf(%s)' % type.to_s.camelize
          },
  
          'oneOf' => ->(*options) {
            enums = options.map{ |k| "'#{k.to_s}'" }.join(',')
            'PropTypes.oneOf([%s])' % enums
          },
  
          'oneOfType' => ->(*options) {
            types = options.map{ |k| "#{lookup(k.to_s, k.to_s)}" }.join(',')
            'PropTypes.oneOfType([%s])' % types
          }
        }
  
        TYPESCRIPT_TYPES = {
          'node' =>        'React.ReactNode',
          'bool' =>        'boolean',
          'boolean' =>     'boolean',
          'string' =>      'string',
          'number' =>      'number',
          'object' =>      'object',
          'array' =>       'Array<any>',
          'shape' =>       'object',
          'element' =>     'object',
          'func' =>        'object',
          'function' =>    'object',
          'any' =>         'any',
  
          'instanceOf' => ->(type) {
            type.to_s.camelize
          },
  
          'oneOf' => ->(*opts) {
            opts.map{ |k| "'#{k.to_s}'" }.join(" | ")
          },
  
          'oneOfType' => ->(*opts) {
            opts.map{ |k| "#{ts_lookup(k.to_s, k.to_s)}" }.join(" | ")
          }
        }
  
        def create_component_file
          template_extension = if options[:coffee]
            'js.jsx.coffee'
          elsif options[:ts]
            'js.jsx.tsx.erb'
          elsif options[:es6] || webpacker?
            'es6.jsx'
          else
            'js.jsx'
          end
  
          # Prefer webpacker to sprockets:
          if webpacker?
            new_file_name = file_name.camelize
            extension = if options[:coffee]
              'coffee'
            elsif options[:ts]
              'tsx'
            else
              'js'
            end
            target_dir = webpack_configuration.source_path
              .join('components')
              .relative_path_from(::Rails.root)
              .to_s
          else
            new_file_name = file_name
            extension = template_extension
            target_dir = 'app/assets/javascripts/components'
          end
  
          file_path = File.join(target_dir, class_path, "#{new_file_name}.#{extension}")
          template("component.#{template_extension}", file_path)
        end
  
        private
  
        def webpack_configuration
          Webpacker.respond_to?(:config) ? Webpacker.config : Webpacker::Configuration
        end
  
        def component_name
          file_name.camelize
        end
  
        def file_header
          if webpacker?
            return %|import * as React from "react"\n| if options[:ts]
            %|import React from "react"\nimport PropTypes from "prop-types"\n|
          else
            ''
          end
        end
  
        def file_footer
          if webpacker?
            %|export default #{component_name}|
          else
            ''
          end
        end
  
        def webpacker?
          defined?(Webpacker)
        end
  
        def parse_attributes!
          self.attributes = (attributes || []).map do |attr|
            name = ''
            type = ''
            args = ''
            args_regex = /(?<args>{.*})/
  
            name, type = attr.split(':')
  
            if matchdata = args_regex.match(type)
              args = matchdata[:args]
              type = type.gsub(args_regex, '')
            end
  
            if options[:ts]
              { :name => name, :type => ts_lookup(name, type, args), :union => union?(args) }
            else
              { :name => name, :type => lookup(type, args) }
            end
          end
        end
  
        def union?(args = '')
          return args.to_s.gsub(/[{}]/, '').split(',').count > 1
        end
  
        def self.ts_lookup(name, type = 'node', args = '')
          ts_type = TYPESCRIPT_TYPES[type]
          if ts_type.blank?
            if type =~ /^[[:upper:]]/
              ts_type = TYPESCRIPT_TYPES['instanceOf']
            else
              ts_type = TYPESCRIPT_TYPES['node']
            end
          end
  
          args = args.to_s.gsub(/[{}]/, '').split(',')
  
          if ts_type.respond_to? :call
            if args.blank?
              return ts_type.call(type)
            end
  
            ts_type = ts_type.call(*args)
          end
  
          ts_type
        end
  
        def ts_lookup(name, type = 'node', args = '')
          self.class.ts_lookup(name, type, args)
        end
  
         def self.lookup(type = 'node', options = '')
           react_prop_type = REACT_PROP_TYPES[type]
           if react_prop_type.blank?
             if type =~ /^[[:upper:]]/
               react_prop_type = REACT_PROP_TYPES['instanceOf']
             else
               react_prop_type = REACT_PROP_TYPES['node']
             end
           end
  
           options = options.to_s.gsub(/[{}]/, '').split(',')
  
           react_prop_type = react_prop_type.call(*options) if react_prop_type.respond_to? :call
           react_prop_type
         end
  
         def lookup(type = 'node', options = '')
           self.class.lookup(type, options)
         end
      end
    end
  end