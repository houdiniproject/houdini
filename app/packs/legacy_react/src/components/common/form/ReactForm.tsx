// License: LGPL-3.0-or-later
import * as React from 'react';
import { observer, Provider } from 'mobx-react';
import {Field, Form} from "mobx-react-form";
import {observable, action, toJS} from 'mobx';

export interface ReactFormProps
{
  form:Form
}

///Mostly useless class but, at some point, will replace all our form elements
@observer
export class ReactForm extends React.Component<ReactFormProps, {}> {



  @observable
  form:Form

  @action.bound
  componentWillMount()
  {
    this.form = this.props.form
  }




  componentDidUpdate(){


  }

  componentWillUnmount(){
  }


  render() {

     return <form >
          {this.props.children}
      </form>
  }
}




