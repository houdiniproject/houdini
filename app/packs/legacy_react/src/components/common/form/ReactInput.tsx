// License: LGPL-3.0-or-later
import  * as React from 'react';
import { observer, inject, Provider } from 'mobx-react';
import {injectIntl} from 'react-intl';
import {Field} from "mobx-react-form";
import {observable, action, toJS, runInAction} from 'mobx';
import {InputHTMLAttributes} from 'react';
import {ReactInputProps} from "./react_input_props";
import {SelectHTMLAttributes} from "react";
import {ReactSelectProps} from "./ReactSelect";
import {castToNullIfUndef} from "../../../lib/utils";


type InputTypes = ReactInputProps &
  InputHTMLAttributes<HTMLInputElement>

class ReactInput extends React.Component<InputTypes, {}> {

  constructor(props:InputTypes){
    super(props)
  }

  @observable
  field:Field


  @action.bound
  componentWillMount(){

    this.field = this.props.field


    this.updateProps()
  }

  componentWillUnmount(){
  }


  componentDidUpdate(prevProps: Readonly<InputTypes>, prevState: Readonly<{}>): void {
    this.updateProps()
  }

  @action.bound
  updateProps() {
      this.field.set('label', castToNullIfUndef(this.props.label))
      this.field.set('placeholder', castToNullIfUndef(this.props.placeholder))
  }

  ///Removes the properties we don't want to put into the input element
  @action.bound
  winnowProps(): InputTypes {
    let ourProps = {...this.props}
    delete ourProps.field
    delete ourProps.value
    return ourProps

  }

  render() {
      return <input {...this.winnowProps()} {...this.field.bind()}/>
  }
}

export default observer(ReactInput)



