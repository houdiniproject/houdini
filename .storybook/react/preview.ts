// License: LGPL-3.0-or-later
import {decorators} from './preview_common';

require('../../app/javascript/components/tests/decorators/roboto')

export {decorators};
