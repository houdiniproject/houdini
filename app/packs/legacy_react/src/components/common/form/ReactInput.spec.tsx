// License: LGPL-3.0-or-later
import * as React from 'react';
import 'jest';
import ReactInput from './ReactInput'
import {Form} from "mobx-react-form";
import {mount} from 'enzyme';
import {toJS, observable, action, runInAction} from 'mobx';
import {observer} from 'mobx-react';
import {ReactForm} from "./ReactForm";


@observer
class TestChange extends React.Component {
  @observable
  remove: boolean
  @observable
  form: Form

  @action.bound
  componentWillMount() {
    this.form = new Form({
      fields: [{
        name: 'name',
        extra: null
      }
      ]
    })
  }


  @action.bound
  onClick() {
    this.remove = true
  }

  render() {
    let reactInput = !this.remove ? <ReactInput field={this.form.$('name')} label={'label1'} placeholder={"holder"}>

    </ReactInput> : undefined

    return <ReactForm form={this.form}>

      {reactInput}
      <button onClick={() => this.onClick()}/>
    </ReactForm>
  }
}

describe('ReactInput', () => {

  let form: Form
  beforeEach(() => {
    form = new Form({
      fields: [
        {
          name: 'name',
          extra: null
        }
      ]
    })
  })

  test('gets added properly', () => {
    let res = mount(<ReactForm form={form}>
      <ReactInput field={form.$('name')} label={"label"}
                  placeholder={"holder"} value={'snapshot'} aria-required={true}/>

    </ReactForm>)


    //Did the attributes settings work as expected back to the objects
    expect(form.$('name').label).toEqual('label')
    expect(form.$('name').placeholder).toEqual('holder')
    expect(form.$('name').value).toEqual('')

    //is the aria attribute passted through to the input
    let input = res.find('input')
    expect(input.prop('aria-required')).toEqual(true)


    // is the input properly bound?
    input.simulate('change', {target: {value: 'something'}})
    expect(form.$('name').value).toEqual('something')
  })

  test('gets removed properly', () => {

    let res = mount(<TestChange/>)

    // The two casts are needed because Typescript was going blowing up without the 'any' first.
    // Why was it? *shrugs*
    let f = res.find('ReactForm').instance() as any as ReactForm
    expect(f.form.size).toEqual(1)

    res.find('input').simulate('change', {target: {value: 'something'}})

    expect(f.form.$('name').value).toEqual('something')

    res.find('button').simulate('click')
    expect(f.form.size).toEqual(1)

    expect(toJS(res.find('form'))).toMatchSnapshot()

    expect(f.form.$('name').label).toEqual('label1')
    expect(f.form.$('name').placeholder).toEqual('holder')
  })
})