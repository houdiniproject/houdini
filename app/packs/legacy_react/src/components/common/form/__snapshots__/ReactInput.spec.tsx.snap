// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactInput gets removed properly 1`] = `
<form>
  <button
    onClick={[Function]}
  />
</form>
`;
